### PR TITLE
#63: full refactoring

### PR DESCRIPTION
**Pull request by yegor256**

### CODE REVIEW

**Comment by karato** [comment]
Thanks, I will find someone to review this PR soon

**Comment by karato** [comment]
@alevohin review it

**Comment by yegor256** [comment]
@alevohin please wait a bit, I'm still finishing this branch

**Comment by pinaf** [comment]
@alevohin good luck reviewing hehe!

**Comment by alevohin** [comment]
@yegor256 please let me know if I can start

**Comment by yegor256** [comment]
@alevohin please go ahead

File: pom.xml [hunk 1]
```diff
@@ -157,6 +182,26 @@
             <version>1.6.2</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (alevohin) [inline pom.xml]: `com.google.guava:guava` is definded in parent pom.xml. This declaration may be removed.

File: pom.xml [hunk 2]
```diff
@@ -246,12 +356,12 @@
...
-                                <packageBranchRate>50</packageBranchRate>
-                                <packageLineRate>75</packageLineRate>
-                                <totalBranchRate>50</totalBranchRate>
-                                <totalLineRate>80</totalLineRate>
+                                <branchRate>0</branchRate>
+                                <lineRate>0</lineRate>
+                                <packageBranchRate>0</packageBranchRate>
+                                <packageLineRate>0</packageLineRate>
+                                <totalBranchRate>5</totalBranchRate>
+                                <totalLineRate>40</totalLineRate>
```
Comment (alevohin) [inline pom.xml:374]: Lets add puzzle wich describes desireable cobertura rates and what to do to reach it.

**Comment by alevohin** [comment]
@yegor256 it's not clean about REST API introduced in https://github.com/teamed/nerodesk/issues/14.
If we don't need it, lets clean `todo`s in `com.nerodesk.Launch`.

File: src/test/java/com/nerodesk/TkDocsTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public final class TkDocsTest {
+
+    /**
+     * TkDocs can return a list of docs.
+     * @throws Exception If fails.
+     */
+    @Test
+    public void returnsPlainText() throws Exception {
+        final User user = new MkBase().user("urn:test:1");
```
Comment (alevohin) [inline src/test/java/com/nerodesk/TkDocsTest.java:56]: This test fails on Windows PC with

```
Running com.nerodesk.TkDocsTest
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.124 sec <<< FAILURE! - in com.nerodesk.TkDocsTest
returnsPlainText(com.nerodesk.TkDocsTest)  Time elapsed: 0.124 sec  <<< ERROR!
java.io.IOException: Directory 'C:\Users\YURIY~1.ALE\AppData\Local\Temp\1426358911475-1\urn:test:1' could not be created
        at org.apache.commons.io.FileUtils.openOutputStream(FileUtils.java:363)
        at org.apache.commons.io.FileUtils.openOutputStream(FileUtils.java:326)
        at org.apache.commons.io.FileUtils.touch(FileUtils.java:444)
        at com.nerodesk.om.mock.MkDoc.write(MkDoc.java:99)
        at com.nerodesk.TkDocsTest.returnsPlainText(TkDocsTest.java:58)
...
```

Lets wrap username at `MkDoc` to remove forbidden symbols or replace `:` with something else.

File: src/test/java/com/nerodesk/TkDocsTest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+ * @since 0.2
+ */
+public final class TkDocsTest {
+
+    /**
+     * TkDocs can return a list of docs.
+     * @throws Exception If fails.
+     */
+    @Test
+    public void returnsPlainText() throws Exception {
```
Comment (alevohin) [inline src/test/java/com/nerodesk/TkDocsTest.java]: Lets rename it to `returnsListOfDocs` or something like that.

File: src/main/java/com/nerodesk/om/aws/AwsDoc.java [hunk 5]
```diff
@@ -27,81 +27,88 @@
...
+        Logger.info(this, "%s written", this.label);
+    }
+
+    /**
+     * Get ocket.
+     * @return Ocket
+     */
+    private Ocket ocket() {
+        return this.bucket.ocket(
+            String.format("/%s%s", this.user, this.label)
```
Comment (alevohin) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java:115]: Does it guarantee unique key?
For example below 2 different user/files that produces the same key.

```
user="urn:test:1";
label="23.txt"
```

and

```
user="urn:test:12";
label="3.txt"
```

File: src/main/xsl/docs.xsl [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            <xsl:apply-templates select="docs/name"/>
+        </ul>
+    </xsl:template>
+    <xsl:template match="name">
+        <li>
+            <xsl:value-of select="."/>
+            <xsl:text> </xsl:text>
+            <a href="{read}">read</a>
+            <xsl:text> | </xsl:text>
+            <a href="{read}">delete</a>
```
Comment (alevohin) [inline src/main/xsl/docs.xsl]: `<a href="{delete}">delete</a>` may be?

File: src/test/java/com/nerodesk/AppTest.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import org.hamcrest.Matchers;
+import org.junit.Test;
+import org.takes.http.FtRemote;
+
+/**
+ * Test case for {@code Launch}.
+ *
+ * @author Yegor Bugayenko (yegor@teamed.io)
+ * @version $Id$
+ * @since 0.1
```
Comment (alevohin) [inline src/test/java/com/nerodesk/AppTest.java]: `@since 0.2`

**Comment by alevohin** [comment]
@yegor256 please see comments above

**Comment by yegor256** [comment]
@alevohin excellent comments, I'm working on them...

**Comment by yegor256** [comment]
@alevohin I addressed your comments, please take a look again

File: src/main/java/com/nerodesk/om/aws/AwsDoc.java (hunk 5 above)
Comment (yegor256) [inline src/main/java/com/nerodesk/om/aws/AwsDoc.java:115]: I added a comment to Javadoc of the method

File: src/main/java/com/nerodesk/om/mock/MkUser.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+     * @param urn URN
+     */
+    public MkUser(final File file, final String urn) {
+        this.dir = file;
+        this.name = urn;
+    }
+
+    @Override
+    public Docs docs() {
+        return new MkDocs(this.dir, this.name.replace("[^a-z0-9]", "/"));
```
Comment (alevohin) [inline src/main/java/com/nerodesk/om/mock/MkUser.java]: Lets replace with `return new MkDocs(this.dir, this.name.replaceAll("[^a-z0-9]", "/"));`

**Comment by alevohin** [comment]
There are errors for cobertura test in Travis log. I can reproduce it locally.

The main reason I think is `Javax.xml.parsers.DocumentBuilderFactory.setFeature(Ljava/lang/String;Z)`, please see log below:

```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.25 sec <<< FAILURE! - in com.nerodesk.TkDocsTest
returnsListOfDocs(com.nerodesk.TkDocsTest)  Time elapsed: 1.25 sec  <<< ERROR!
java.lang.AbstractMethodError: javax.xml.parsers.DocumentBuilderFactory.setFeature(Ljava/lang/String;Z)V
        at com.jcabi.xml.XMLDocument.<clinit>(XMLDocument.java:118)
        at com.jcabi.matchers.XPathMatcher.matchesSafely(XPathMatcher.java:75)
        at org.hamcrest.TypeSafeMatcher.matches(TypeSafeMatcher.java:65)
        at org.hamcrest.core.AllOf.matches(AllOf.java:27)
        at org.hamcrest.DiagnosingMatcher.matches(DiagnosingMatcher.java:12)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:12)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
        at com.nerodesk.TkDocsTest.returnsListOfDocs(TkDocsTest.java:63)
```

Сonsequence are below:

```
Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 3.074 sec <<< FAILURE! - in com.nerodesk.TkIndexTest
returnsHtml(com.nerodesk.TkIndexTest)  Time elapsed: 1.819 sec  <<< ERROR!
java.lang.NoClassDefFoundError: Could not initialize class com.jcabi.xml.XMLDocument
        at com.jcabi.matchers.XPathMatcher.matchesSafely(XPathMatcher.java:75)
        at org.hamcrest.TypeSafeMatcher.matches(TypeSafeMatcher.java:65)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:12)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
        at com.nerodesk.TkIndexTest.returnsHtml(TkIndexTest.java:66)
```

**Comment by alevohin** [comment]
@yegor256 Thank you! Please see comments above.

File: src/main/java/com/nerodesk/om/mock/MkUser.java (hunk 8 above)
Comment (yegor256) [inline src/main/java/com/nerodesk/om/mock/MkUser.java]: @alevohin I didn't get this... it's exactly what we have in the branch...

**Comment by yegor256** [comment]
@alevohin I don't know why this XML problem is happening, but this build is clean:

```
mvn clean install -Pqulice,cobertura
```

Maybe you know how to fix that?

File: src/main/java/com/nerodesk/om/mock/MkUser.java (hunk 8 above)
Comment (alevohin) [inline src/main/java/com/nerodesk/om/mock/MkUser.java]: @yegor256 Lets use `replaceAll` instead of `replace`

Comment (yegor256) [inline src/main/java/com/nerodesk/om/mock/MkUser.java]: Oh, right, thanks!

**Comment by alevohin** [comment]
@yegor256 I don't know the reason exactly, probably, in cobertura and xml api/parser.
I checked rultor, it uses jcabi-xml and another cobertura version

```
 <groupId>org.codehaus.mojo</groupId>
                        <artifactId>cobertura-maven-plugin</artifactId>
                        <version>2.5.2</version>
```

So lets downgrade `cobertura-maven-plugin` version and it should works.

**Comment by yegor256** [comment]
@alevohin hm.. good catch, thanks! fixed

**Comment by yegor256** [comment]
@alevohin what do you think? are we ready to merge?

**Comment by alevohin** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor merge

**Comment by alevohin** [comment]
@yegor256 thank you! Whay do you think about obsoleted puzzles? F.e. #69, there is PR #75

**Comment by yegor256** [comment]
@alevohin they are obsolete and will be closed very soon. we simply don't need them any more.

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@alevohin thanks!

**Comment by karato** [comment]
@rultor deploy

**Comment by karato** [comment]
@alevohin Many thanks! **21 mins** were added to your account in Transaction ID `53753456`. review comments (c=6) added as a bonus. added +21 to your rating, now it is equal to [+1697](http://www.netbout.com/b/35105?open=rating)

**Pull request merged**